File: MoPub/SASMoPubInterstitialAdapter.m
//
//  SASMoPubInterstitialAdapter.m
//  SmartAdServer
//
//  Created by Loïc GIRON DIT METAZ on 25/09/2018.
//  Copyright © 2018 Smart AdServer. All rights reserved.
//

#import "SASMoPubInterstitialAdapter.h"
#import <MoPubSDK/MoPub.h>

NS_ASSUME_NONNULL_BEGIN

@interface SASMoPubInterstitialAdapter () <MPInterstitialAdControllerDelegate>

@property (nonatomic, strong) NSDictionary *clientParameters;

@end

@implementation SASMoPubInterstitialAdapter

#pragma mark - Mediation adapter protocol implementation

- (instancetype)initWithDelegate:(id<SASMediationInterstitialAdapterDelegate>)delegate {
    if (self = [super init]) {
        self.delegate = delegate;
    }
    return self;
}

- (void)requestInterstitialWithServerParameterString:(NSString *)serverParameterString clientParameters:(NSDictionary *)clientParameters {
    self.isReady = NO;
    self.clientParameters = clientParameters;
    
    // Configuring Application ID is done in the base class
    [self configureApplicationIDWithServerParameterString:serverParameterString];
    
    // Creating the MoPub interstitia controller
    self.interstitialController = [MPInterstitialAdController interstitialAdControllerForAdUnitId:self.adUnitID];
    self.interstitialController.delegate = self;
    
    // Loading the interstitial after initialization
    __weak typeof(self) weakSelf = self;
    [self initializeMoPubSDK:^{
        [weakSelf.interstitialController loadAd];
    }];
}

- (void)showInterstitialFromViewController:(UIViewController *)viewController {
    // Configuring GDPR status is done in the base class
    if ([self configureGDPRWithClientParameters:self.clientParameters viewController:viewController]) {
        // If MoPub is attempting to display a CMP consent dialog, we abort the ad show so we don't display an interstitial
        // and the dialog at the same time.
        
        // Call failToShow delegate to reset the interstitial for the next call.
        NSError *error = [NSError errorWithDomain:SASMoPubAdapterErrorDomain
                                             code:SASMoPubAdapterErrorCodeCMPDisplayed
                                         userInfo:@{ NSLocalizedDescriptionKey: @"The MoPub CMP was displayed instead of the interstitial ad." }];
        [self.delegate mediationInterstitialAdapter:self didFailToShowWithError:error];
        
        return;
    }
    
    printf("the interstitial is shown");
    [self.interstitialController showFromViewController:viewController];
}

- (BOOL)isInterstitialReady {
    return self.isReady;
}

#pragma mark - MoPub banner delegate

- (void)interstitialDidLoadAd:(MPInterstitialAdController *)interstitial {
    self.isReady = YES;
    [self.delegate mediationInterstitialAdapterDidLoad:self];
}

- (void)interstitialDidFailToLoadAd:(MPInterstitialAdController *)interstitial withError:(NSError *)error {
    [self.delegate mediationInterstitialAdapter:self didFailToLoadWithError:error noFill:YES];
    // Since there is no documented way to know if the error is due to a 'no fill', we send YES for this parameter
}

- (void)interstitialDidAppear:(MPInterstitialAdController *)interstitial {
    [self.delegate mediationInterstitialAdapterDidShow:self];
}

- (void)interstitialDidDismiss:(MPInterstitialAdController *)interstitial {
    [self.delegate mediationInterstitialAdapterDidClose:self];
}

- (void)interstitialDidExpire:(MPInterstitialAdController *)interstitial {
    // The interstitial is not ready anymore if expired
    self.isReady = NO;
}

- (void)interstitialDidReceiveTapEvent:(MPInterstitialAdController *)interstitial {
    [self.delegate mediationInterstitialAdapterDidReceiveAdClickedEvent:self];
}

@end

NS_ASSUME_NONNULL_END

File: MoPub/SASMoPubInterstitialAdapter.swift
//
//  SASMoPubInterstitialAdapter.swift
//  SmartAdServer
//
//  Created by Loïc GIRON DIT METAZ on 11/08/2021.
//  Copyright © 2021 Smart AdServer. All rights reserved.
//

import UIKit
import SASDisplayKit
import MoPubSDK

/**
 MoPub interstitial adapter class for Smart SDK mediation.
 
 Mediation adapter classes can be used to display an ad using a third party SDK directly from
 an insertion handled by Smart.
 
 To use an adapter class, you simply have to add them to your Xcode project and they will
 be automatically instantiated by the Smart SDK if needed.
 */
@objc(SASMoPubInterstitialAdapter)
class SASMoPubInterstitialAdapter : SASMoPubBaseAdapter, SASMediationInterstitialAdapter {
    
    /// A delegate that this adapter must call to provide information about the ad loading status or events to the Smart SDK.
    private weak var delegate: SASMediationInterstitialAdapterDelegate? = nil
    
    /// The currently displayed MoPub interstitial view if any.
    private var interstitialController: MPInterstitialAdController? = nil
    
    /// YES if an interstitial is ready to be displayed, NO otherwise.
    private var isReady: Bool = false
    
    // Client parameters set by the SDK for the current ad request.
    private var clientParameters: [AnyHashable : Any] = [:]
    
    required init(delegate: SASMediationInterstitialAdapterDelegate) {
        self.delegate = delegate
        super.init()
    }
    
    func requestInterstitial(withServerParameterString serverParameterString: String, clientParameters: [AnyHashable : Any]) {
        self.isReady = false
        self.clientParameters = clientParameters
        
        // Configuring Application ID is done in the base class
        configureApplicationID(serverParameterString: serverParameterString)
        
        // Creating the MoPub interstitial controller
        interstitialController = MPInterstitialAdController(forAdUnitId: adUnitID)
        interstitialController?.delegate = self
        
        // Loading the interstitial after initialization
        initializeMoPubSDK {
            self.interstitialController?.loadAd()
        }
    }
    
    func showInterstitial(from viewController: UIViewController) {
        // Configuring GDPR status is done in the base class
        if configureGDPR(clientParameters: self.clientParameters, viewController: viewController) {
            // If MoPub is attempting to display a CMP consent dialog, we abort the ad show so we don't display an interstitial
            // and the dialog at the same time.
            
            // Call failToShow delegate to reset the interstitial for the next call.
            let error = NSError(
                domain: ErrorConstants.errorDomain,
                code: ErrorConstants.errorCodeCMPDisplayed,
                userInfo: [NSLocalizedDescriptionKey: "The MoPub CMP was displayed instead of the interstitial ad."]
            )
            delegate?.mediationInterstitialAdapter(self, didFailToShowWithError: error)
            
        } else {
            interstitialController?.show(from: viewController)
        }
    }
    
    func isInterstitialReady() -> Bool {
        return self.isReady
    }
    
}

/**
 MoPub delegate implementation.
 */
extension SASMoPubInterstitialAdapter : MPInterstitialAdControllerDelegate {
    
    func interstitialDidLoadAd(_ interstitial: MPInterstitialAdController!) {
        isReady = true
        delegate?.mediationInterstitialAdapterDidLoad(self)
    }
    
    func interstitialDidFail(toLoadAd interstitial: MPInterstitialAdController!, withError error: Error!) {
        delegate?.mediationInterstitialAdapter(self, didFailToLoadWithError: error, noFill: true)
        // Since there is no documented way to know if the error is due to a 'no fill', we send true for this parameter
    }
    
    func interstitialDidAppear(_ interstitial: MPInterstitialAdController!) {
        delegate?.mediationInterstitialAdapterDidShow(self)
    }
    
    func interstitialDidDismiss(_ interstitial: MPInterstitialAdController!) {
        delegate?.mediationInterstitialAdapterDidClose(self)
    }
    
    func interstitialDidExpire(_ interstitial: MPInterstitialAdController!) {
        // The interstitial is not ready anymore if expired
        isReady = false
    }
    
    func interstitialDidReceiveTapEvent(_ interstitial: MPInterstitialAdController!) {
        delegate?.mediationInterstitialAdapterDidReceiveAdClickedEvent(self)
    }
    
}
